receiver/hostmetricsreceiver/internal/scraper/processesscraper/processes_scraper.go

```diff
@@ -16,28 +16,64 @@ package processesscraper
 
 import (
 	"context"
+	"sort"
+	"time"
 
 	"github.com/shirou/gopsutil/host"
 	"github.com/shirou/gopsutil/load"
+	"github.com/shirou/gopsutil/process"
 
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/hostmetricsreceiver/internal/metadata"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
 )
 
+var metricsLength = func() int {
+	n := 0
+	if enableProcessesCount {
+		n++
+	}
+	if enableProcessesCreated {
+		n++
+	}
+	return n
+}()
+
 // scraper for Processes Metrics
 type scraper struct {
 	config    *Config
 	startTime pdata.Timestamp
 
-	// for mocking gopsutil load.Misc
-	misc getMiscStats
+	// for mocking gopsutil
+	getMiscStats func() (*load.MiscStat, error)
+	getProcesses func() ([]proc, error)
 }
 
-type getMiscStats func() (*load.MiscStat, error)
+// for mocking out gopsutil process.Process
+type proc interface {
+	Status() (string, error)
+}
+
+type processesMetadata struct {
+	countByStatus    map[string]int64 // ignored if enableProcessesCount is false
+	processesCreated *int64           // ignored if enableProcessesCreated is false
+}
 
 // newProcessesScraper creates a set of Processes related metrics
 func newProcessesScraper(_ context.Context, cfg *Config) *scraper {
-	return &scraper{config: cfg, misc: load.Misc}
+	return &scraper{
+		config:       cfg,
+		getMiscStats: load.Misc,
+		getProcesses: func() ([]proc, error) {
+			ps, err := process.Processes()
+			ret := make([]proc, len(ps))
```

Comment:
why do we need to copy the processes list here?

Reply:
We need to convert a slice of structs to a slice of interfaces. As they have a different memory layout, we can't just cast.

We need to convert to an interface so that we can mock out getProcesses in testing.

Reply:
got it